File: app/src/main/java/com/reddittask/android/ui/activity/main/MainActivity.java
package com.reddittask.android.ui.activity.main;

import android.net.Uri;
import android.os.Bundle;
import android.support.customtabs.CustomTabsIntent;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.reddittask.android.R;
import com.reddittask.android.api.Queries;
import com.reddittask.android.model.Child;
import com.reddittask.android.model.GameData;
import com.reddittask.android.ui.activity.base.BaseActivity;
import com.reddittask.android.ui.adapter.RecyclerNewsDataAdapter;
import com.reddittask.android.ui.dialog.MaterialDialog;
import com.reddittask.android.ui.listener.PaginationScrollListener;
import com.reddittask.android.util.CommonUtil;
import com.reddittask.android.util.UiUtil;

import java.util.List;

import butterknife.BindView;

import static com.reddittask.android.Const.PAGE_START;
import static com.reddittask.android.Const.TOTAL_PAGES;

public class MainActivity extends BaseActivity implements MainView {

    @BindView(R.id.recycler_movie)
    protected RecyclerView mRecyclerViewNews;
    private LinearLayoutManager mLinearLayoutManager;
    private RecyclerNewsDataAdapter mNewsDataAdapter;
    private MaterialDialog mProgressBar;
    private MainPresenterImpl mMainPresenter;
    private boolean mIsLoading = false;
    private boolean mIsLastPage = false;
    private int mCurrentPage = PAGE_START;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mMainPresenter = new MainPresenterImpl(this, new Queries(), CommonUtil.convertStringToInt(PAGE_START));

        init();
        mMainPresenter.onAttach(this);
        addLoadMoreListener();
    }

    @Override
    public int getLayout() {
        return R.layout.activity_main;
    }

    private void init() {
        mProgressBar = new MaterialDialog(this);
        mLinearLayoutManager = new LinearLayoutManager(this);
        mRecyclerViewNews.setLayoutManager(mLinearLayoutManager);
    }

    private void addLoadMoreListener() {
        mRecyclerViewNews.addOnScrollListener(new PaginationScrollListener(mLinearLayoutManager) {
            @Override
            protected void loadMoreItems() {
                mIsLoading = true;
                mCurrentPage += 1;
                mMainPresenter.loadNextData(CommonUtil.convertStringToInt(mCurrentPage));
            }


            @Override
            public int getTotalPageCount() {
                return TOTAL_PAGES;
            }

            @Override
            public boolean isLastPage() {
                return mIsLastPage;
            }

            @Override
            public boolean isLoading() {
                return mIsLoading;
            }
        });
    }

    @Override
    public void showLoading() {
        mProgressBar.show();
    }

    @Override
    public void hideLoading() {
        mProgressBar.hide();
    }

    @Override
    public void openDetailGame(int position, String url) {
        // Build the custom tab intent and launch the url
        CustomTabsIntent customTabsIntent = UiUtil.buildCustomTabsIntent(this);
        customTabsIntent.launchUrl(MainActivity.this, Uri.parse(url));
    }

    @Override
    public void displayGameData(GameData gameData) {
        if (mNewsDataAdapter == null) {
            mNewsDataAdapter = new RecyclerNewsDataAdapter(gameData, this, mMainPresenter);
            mRecyclerViewNews.setAdapter(mNewsDataAdapter);
        } else {
            mNewsDataAdapter.removeLoadingFooter();
        }
        List<Child> gameDataChildren = getLastGames(mNewsDataAdapter.getItemCount(), gameData.getData().getChildren());
        mIsLoading = false;

        mNewsDataAdapter.addAll(gameDataChildren);

        if (mCurrentPage != TOTAL_PAGES) mNewsDataAdapter.addLoadingFooter();
        else mIsLastPage = true;
    }

    @Override
    public void showErrorMessage() {
        Toast.makeText(this, getResources().getString(R.string.error_msg), Toast.LENGTH_SHORT).show();
    }

    public static List<Child> getLastGames(int itemCount, List<Child> allListData) {
        return allListData.subList(itemCount, allListData.size());
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mMainPresenter.onDetach();
    }
}

File: app/src/main/java/com/keddits/ui/activity/main/MainActivity.kt
package com.keddits.ui.activity.main

import android.net.Uri
import android.os.Bundle
import android.support.customtabs.CustomTabsIntent
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.widget.Toast
import com.keddits.Const
import com.keddits.R
import com.keddits.api.Queries
import com.keddits.model.Child
import com.keddits.ui.activity.base.BaseActivity
import com.keddits.ui.adapter.RecyclerNewsDataAdapter
import com.keddits.ui.dialog.MaterialDialog
import com.keddits.ui.listener.PaginationScrollListener
import com.keddits.util.UiUtil
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main.view.*

/**
 * Created by marinaracu on 21.12.2017.
 */
class MainActivity : BaseActivity(), MainView {

    private var isLoading: Boolean = false
    private var isLastPage: Boolean = false
    private var currentPage: Int = Const.PAGE_START
    private lateinit var mainPresenter: MainPresenterImpl
    private lateinit var matiralProgress: MaterialDialog
    private lateinit var linearLayoutManager: LinearLayoutManager
    private lateinit var recyclerAdapter: RecyclerNewsDataAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mainPresenter = MainPresenterImpl(Queries(), Const.PAGE_START)
        init()
        mainPresenter.onAttach(this)
        addLoadMoreListener()
    }

    override fun getLayout(): Int {
        return R.layout.activity_main
    }

    private fun init() {
        matiralProgress = MaterialDialog(this)
        linearLayoutManager = LinearLayoutManager(this)
        recycler_movie.layoutManager = linearLayoutManager
        recyclerAdapter = RecyclerNewsDataAdapter { mainPresenter.onItemNewsClicked(it.data!!.url) }
        recycler_movie.adapter = recyclerAdapter
    }

    private fun addLoadMoreListener() {
        recycler_movie.addOnScrollListener(object : PaginationScrollListener(linearLayoutManager) {
            override fun loadMoreItems() {
                isLoading = true
                currentPage += 10
                mainPresenter.loadNextData(currentPage, recyclerAdapter.itemCount)
            }

            override fun getTotalPageCount(): Int {
                return Const.TOTAL_PAGES
            }

            override fun isLastPage(): Boolean {
                return isLastPage
            }

            override fun isLoading(): Boolean {
                return isLoading
            }

        })
    }

    override fun showLoadding() {
        matiralProgress.show()
    }

    override fun hideLoadding() {
        matiralProgress.hide()
    }

    override fun openDetailGame(url: String) {
        // Build the custom tab intent and launch the url
        val customTabIntent: CustomTabsIntent = UiUtil.buildCustomTabsIntent(this)
        customTabIntent.launchUrl(this, Uri.parse(url))
    }

    override fun displayGameData(gameData: MutableList<Child>) {
        Log.d("DebugH-0",  currentPage.toString())
        if (recyclerAdapter.itemCount == 0) {
            recyclerAdapter.setGameData(gameData)
        } else {
            recyclerAdapter.removeLoadingFooter()
        }
        Log.d("DebugH-1",  currentPage.toString())
        mainPresenter.loadLastData(gameData, recyclerAdapter.itemCount)
    }

    override fun showErrorMessage() {
        Toast.makeText(this, R.string.error_msg, Toast.LENGTH_SHORT).show()
    }

    override fun showLastItemLists(cildren: MutableList<Child>) {
        isLoading = false
        recyclerAdapter.addAll(cildren)

        if (currentPage != Const.TOTAL_PAGES) recyclerAdapter.addLoadingFooter()
        else isLastPage = true
    }

    override fun onDestroy() {
        super.onDestroy()
        mainPresenter.onDetach()
    }
}